### ISSUE TITLE
get_fills documentation out of date

### ISSUE THREAD

**Issue by noah222** [danpaquin/coinbasepro-python#358]
The example for paginated get fills using 'before' command does not work because it does not include the product type that is now required.

The following line in the example:

`new_fills = auth_client.get_fills(before=recent_fills[0]['trade_id'])`

Should  be updated to:

`new_fills = auth_client.get_fills(before=recent_fills[0]['trade_id'], product_id="***-USD")`

This of course requires the use of the correct product_id to match up with the trade you are looking up ("ETH-USD", "BTC-USD" etc.)

**Comment by nazariyv** [comment]
Noah, looks like this library isn't maintained anymore. Would you be keen to fork it and me + my friend @TristanJM would be happy to help to maintain it, as we are using it too?

**Comment by acontry** [comment]
FYI I am developing a fork of this project at https://github.com/acontry/coinbasepro where this problem is already fixed.

**Comment by nazariyv** [comment]
@acontry I can't find websocket in there.

**Comment by acontry** [comment]
Initial focus was to build a clean, well-documented REST client which has been stable for a while. Privately I use my own websocket code and FIX client but I haven't cleaned those up for release like the REST stuff.

**Comment by noah222** [comment]
@acontry Might I ask about the error handling of your code? I've got my own clunky error handling that works, but something occasionally slips through. Does it deal with null responses, and timeouts well? Thanks!

**Comment by nazariyv** [comment]
@acontry Would you be able to release the FIX stuff? That would be marvelous.

**Comment by acontry** [comment]
@noah222 The goal is to handle all exceptions with a consistent interface - Ideally any type of exception raises a `CoinbaseAPIError` (or a child of that exception type) with my approach. I just looked, and realized that a timeout will throw a separate exception so I can fix that. Regarding null responses, I do cover cases that I've encountered but of course it's hard to know if I've covered all of them (especially undocumented/off spec stuff).

@nazariyv There is work to be done before releasing FIX, but I'll keep this in mind!

**Comment by nazariyv** [comment]
Well... with the new coinbase fees we can all happily move to binance... 😠

**Comment by noah222** [comment]
the game gets a little harder with these fees, but I've already invested so
much time in developing a working strategy. So I'll be sticking with this
for now. Now it's time to double down and make it even better, reduce my
error rate from 8% to lower than that, if possible.

**Comment by uclatommy** [comment]
If anyone was able to beat a buy and hold strategy in February, I’d love to hear how you did it.

**Comment by nazariyv** [comment]
If anyone did, no one will tell ;)

**Comment by nazariyv** [comment]
What's a good way to get in touch with you Noah?

**Comment by noah222** [comment]
I've put my email in my profile. In theory the trading solution for any time period can be found by backtesting all relevant trading variables and combinations of indicators over that time-frame and using back-propagation to refine the result.  I've built my own testing model to do this that can work for any exchange. The result will be the largest gain using some arbitrary pattern with minimal risk, and typically follows in the footsteps of the market maker. I'm continuously simulating the market and adapting to it. To adapt to these fees I just change the fee variable from 0 to .0015 and then this increases the minimum move that is profitable. The system works away and finds a path, just like a self-driving car or the navigation system of a drone, or in more mythic terms, a biological creature with a survival goal. I can't think of a better way to solve the problem given all the information and tools available. In theory it always works, in practice there is a small gap between having a technical plan and executing it properly in real time. So closing that execution gap and refining the model lead to better and better results as time goes on.

**Comment by uclatommy** [comment]
@noah222, so looks like you are using an ai approach since you mentioned back-propagation? I also have a robust backtesting system where I can swap out any algorithm I want to backtest. And just like you, I'm simulating the cbpro server for backtesting while using historical data to replay to my trading algo. Mine is also able to accommodate the changes to fee structure-- just a matter of changing the fee assumptions. However, unlike you, I'm taking a statistical approach. Namely, a seasonal autoregressive integrated moving average model. It works decently but still was not able to predict the flash crash at the end of February. I'm not yet familiar enough with neural networks to implement my own yet, but eventually, I'm thinking of building a network that can learn how to set the parameters of the seasonal arima model and keep it constantly training on new data. One thing that I have found to help my algo is to build my own type of stop-loss order that can follow prices up or down until it detects an inflection point so even if the algo is a bit early in creating the order, the stop loss can ratchet itself up or down until it finds the peak or trough. Despite all this, I still cannot consistently beat the HODLers.

**Comment by noah222** [comment]
@uclatommy Yes, I would call it brute force combitorial machine learning. This is because my decision nodes are simple and the power is not really in them, but in the technique of sending hundreds of thousands of variations through the simulation to see what comes out the other side. Good ideas are saved and recombined. I have found that trying to predict anything is hard. But reacting quickly makes more sense. I would think selling at the peak with a trailing stop and buying more crypto for less than the sell price would generally be better than holding, but I have to be careful now about over-trading especially with these fees.

**Comment by acontry** [comment]
FYI pushed some improvements over at my repo - also did some more thorough testing of error handling and it behaves quite nicely. See some examples in the readme!

**Comment by nazariyv** [comment]
Isn't that just overfitting? How can you tell it will work out-of-sample? Are you selecting combinations based on the best out-of-sample MSE?

**Comment by noah222** [comment]
not when it works on the minute chart for a year

**Comment by nazariyv** [comment]
@noah222 we need to talk :) I will ping you at some point

**Comment by uclatommy** [comment]
Well, regardless of how everyone is doing with their chosen methodologies, if coinbase achieves their stated goal of increasing liquidity and decreasing volatility, it changes the behavior of the markets. More than the addition of frictional costs on both the taker and maker side of transactions, I would worry about the new market dynamics invalidating most of our backtesting so far. For any currently successful bots, I would probably trade a smaller portfolio when the fee structure changes just to make sure the new dynamics don't suddenly make successful algos unsuccessful. On the bright side, now that the cost differential is smaller between limit and market orders, it might make sense to develop new strategies that can switch between the two types of orders depending on which is more advantageous at the moment. Good luck!
